File: apps/frontend/src/app/api/services.ts
import { apiFetch } from "./fetch";

export const servicesApi = {
  listServices: (parentid?: string) =>
    apiFetch<{
      services: Array<{
        serviceid: string;
        name: string;
        algorithm: number;
        algorithm_label: string;
        sortorder: number;
        weight: number;
        status: number;
        description: string;
        tags: Array<{ tag: string; value: string }>;
        children: Array<{ serviceid: string; name: string }>;
        parents: Array<{ serviceid: string; name: string }>;
      }>;
    }>(`/services${parentid ? `?parentid=${parentid}` : ""}`),
  createService: (payload: {
    name: string;
    algorithm?: number;
    sortorder?: number;
    weight?: number;
    description?: string;
  }) =>
    apiFetch<{ serviceid: string }>("/services", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  updateService: (
    serviceid: string,
    payload: { name?: string; algorithm?: number; description?: string },
  ) =>
    apiFetch<{ ok: boolean }>(`/services/${serviceid}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  deleteService: (serviceid: string) =>
    apiFetch<{ ok: boolean }>(`/services/${serviceid}`, { method: "DELETE" }),
  listSlas: () =>
    apiFetch<{
      slas: Array<{
        slaid: string;
        name: string;
        slo: number;
        period: string;
        period_label: string;
        timezone: string;
        description: string;
        status: number;
        effective_date: number;
        service_tags: Array<{ tag: string; value: string }>;
      }>;
    }>("/sla"),
  createSla: (payload: {
    name: string;
    slo: number;
    period?: string;
    timezone?: string;
    description?: string;
    service_tags?: Array<{ tag: string; value: string }>;
  }) =>
    apiFetch<{ slaid: string }>("/sla", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  deleteSla: (slaid: string) => apiFetch<{ ok: boolean }>(`/sla/${slaid}`, { method: "DELETE" }),
  getSlaReport: (slaid: string, periods?: number) =>
    apiFetch<{ report: Array<Record<string, unknown>> }>(
      `/sla/${slaid}/report${periods ? `?periods=${periods}` : ""}`,
    ),
  listHealthMonitors: (hostid?: string) =>
    apiFetch<{
      monitors: Array<{
        itemid: string;
        name: string;
        host: string;
        hostid: string;
        url: string;
        expected: string;
        running: boolean;
        working: boolean;
        last_value: string | null;
        last_check: number | null;
        proc_itemid: string | null;
        has_proc_check: boolean;
      }>;
    }>(`/health-monitors${hostid ? `?hostid=${hostid}` : ""}`),
  createHealthMonitor: (payload: {
    hostid: string;
    name: string;
    url: string;
    expected_contains?: string;
    process_name?: string;
  }) =>
    apiFetch<{ itemid: string; proc_itemid: string | null }>("/health-monitors", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  deleteHealthMonitor: (itemid: string) =>
    apiFetch<{ ok: boolean }>(`/health-monitors/${itemid}`, { method: "DELETE" }),
};

File: apps/backend/api/routes/services.py
from api.deps import zabbix_call
from fastapi import APIRouter, Depends, Query
from Auth import get_current_user, require_admin, require_operator
from api.managers import services_bot
from api.schemas import (
    HealthMonitorCreateRequest,
    ServiceCreateRequest,
    ServiceUpdateRequest,
    SlaCreateRequest,
)

router = APIRouter(tags=["Services"])


@router.get("/services")
def list_services(parentid: str | None = Query(None), _user=Depends(get_current_user)):
    return {"services": services_bot.list_services(parentid=parentid)}


@router.post("/services")
def create_service(body: ServiceCreateRequest, _user=Depends(require_admin)):
    with zabbix_call():
        sid = services_bot.create_service(
            body.name, body.algorithm, body.sortorder, body.weight, body.description
        )
        return {"serviceid": sid}


@router.put("/services/{serviceid}")
def update_service(
    serviceid: str, body: ServiceUpdateRequest, _user=Depends(require_admin)
):
    with zabbix_call():
        services_bot.update_service(
            serviceid, body.name, body.algorithm, body.description
        )
        return {"ok": True}


@router.delete("/services/{serviceid}")
def delete_service(serviceid: str, _user=Depends(require_admin)):
    with zabbix_call():
        services_bot.delete_service(serviceid)
        return {"ok": True}


@router.get("/sla")
def list_slas(_user=Depends(get_current_user)):
    return {"slas": services_bot.list_slas()}


@router.post("/sla")
def create_sla(body: SlaCreateRequest, _user=Depends(require_admin)):
    with zabbix_call():
        sid = services_bot.create_sla(
            body.name,
            body.slo,
            body.period,
            body.timezone,
            body.description,
            body.service_tags,
        )
        return {"slaid": sid}


@router.delete("/sla/{slaid}")
def delete_sla(slaid: str, _user=Depends(require_admin)):
    with zabbix_call():
        services_bot.delete_sla(slaid)
        return {"ok": True}


@router.get("/sla/{slaid}/report")
def get_sla_report(
    slaid: str, periods: int = Query(3, ge=1, le=12), _user=Depends(get_current_user)
):
    return {"report": services_bot.get_sla_report(slaid, periods)}


@router.get("/health-monitors")
def list_health_monitors(
    hostid: str | None = Query(None), _user=Depends(get_current_user)
):
    return {"monitors": services_bot.list_health_monitors(hostid=hostid)}


@router.post("/health-monitors")
def create_health_monitor(
    body: HealthMonitorCreateRequest, _user=Depends(require_operator)
):
    with zabbix_call():
        return services_bot.add_health_monitor(
            body.hostid, body.name, body.url, body.expected_contains, body.process_name
        )


@router.delete("/health-monitors/{itemid}")
def delete_health_monitor(itemid: str, _user=Depends(require_operator)):
    with zabbix_call():
        services_bot.delete_health_monitor(itemid)
        return {"ok": True}
